// mesh-base/util/qr-reader-source-reader-browser.js
//..............................................................................
const {util} = cxq;
//..............................................................................

//..............................................................................
const {isString} = util;
//..............................................................................

//..............................................................................
function getImageDataFromCanvasEl(canvasEl, callback)
{
    const imageData = canvasEl.getContext('2d').getImageData(0,0,canvasEl.width,canvasEl.height);
    callback(null, imageData);
}
//..............................................................................

//..............................................................................
function getImageDataFromImageEl(imageEl, callback)
{
    const canvasEl = document.createElement('canvas');
    canvasEl.width  = imageEl.width;
    canvasEl.height = imageEl.height;
    
    canvasEl.drawImage(imageEl, 0, 0);

    getImageDataFromCanvasEl(canvasEl, callback);
}
//..............................................................................

//..............................................................................
function getImageDataFromString(source, callback)
{
    var imageEl;
    //..........................................................................
    function callback_onload(evt)
    {
        getImageDataFromImageEl(imageEl, callback);
    }
    //..........................................................................
    function callback_onerror(err)
    {
        callback(err);
    }
    //..........................................................................
    imageEl         = document.createElement('image');
    imageEl.src     = source;
    imageEl.onload  = callback_onload;
    imageEl.onerror = callback_onerror;
}
//..............................................................................

//..............................................................................
function readImage(source, callback)
{    
    var fn = null;

    if (!callback                                      ) return;
    if (!source                                        ) return callback('Source not defined');      
    if (source.data && source.width && source.height   ) return callback(null, source);
    
         if (source instanceof HTMLCanvasElement) fn = getImageDataFromCanvasEl;
    else if (source instanceof HTMLImageElement ) fn = getImageDataFromImageEl;
    else if (isString(source)                   ) fn = getImageDataFromString;

    if (fn)
        fn(source, callback);
    else
        callback('Unsupported source');   
}
//..............................................................................

module.exports = readImage;
