// mesh-base/system-metadata/comparators.js
const {util} = cxq;

//..............................................................................
const {makeColor, makeDate, isDefined, isString} = util;
//..............................................................................

//..............................................................................
function dateToNumber(date)
{
    return date ? makeDate(date).getTime() : 0;
}
//..............................................................................

//..............................................................................
function jsonToString(json)
{
    var str;

    try
    {
        str = json.stringify(json);
    }
    catch(e)
    {
        str = '';
    }

    return str;
}
//..............................................................................

//..............................................................................
function resolveNumber(a,b)
{
    if (!isDefined(a)) a = 0;
    if (!isDefined(b)) b = 0;

    return a-b;
}
//..............................................................................

//..............................................................................
function resolveString(a,b)
{
    if (!isDefined(a)) a = '';
    if (!isDefined(b)) b = '';

    if (!isString(a)) a = String(a);
    if (!isString(b)) b = String(b);

    a = a.toUpperCase();
    b = b.toUpperCase();

    if (a < b) return -1;
    if (a > b) return  1;

    return 0;
}
//..............................................................................

//..............................................................................
function number(a,b)
{
    return resolveNumber(a,b);
}
//..............................................................................

//..............................................................................
function float(a,b)
{
    return resolveNumber(a,b);
}
//..............................................................................

//..............................................................................
function integer(a,b)
{
    return resolveNumber(a,b);
}
//..............................................................................

//..............................................................................
function string(a,b)
{
    return resolveString(a,b);
}
//..............................................................................

//..............................................................................
function json(a,b)
{
    a = jsonToString(a);
    b = jsonToString(b);

    return resolveString(a,b);
}
//..............................................................................

//..............................................................................
function color(a,b)
{
    a = makeColor(a).toHex();
    b = makeColor(b).toHex();

    return resolveString(a,b);
}
//..............................................................................

//..............................................................................
function datetime(a,b)
{
    a = dateToNumber(a);
    b = dateToNumber(b);

    return resolveNumber(a,b);
}
//..............................................................................

//..............................................................................
function boolean(a,b)
{
    if (a === b   ) return  0;
    if (a === true) return -1;

    return 1;
}
//..............................................................................

module.exports = {number, float, integer, string, json, color, datetime, boolean};
