// mesh-base/system-metadata/pin-gpio.js
//..............................................................................
const {util} = cxq;
//..............................................................................

//..............................................................................
const
{
    addToNameSpace,
    processList
} = util;
//..............................................................................

//..............................................................................
const gpio = {};
function buildPinTemplates()
{
    var pinInfo =
    {
        A0      : {index: 0, fn:['PWM' ,'ADC']},
        A1      : {index: 1, fn:['PWM' ,'ADC']},
        A2      : {index: 2, fn:['PWM' ,'ADC']},
        A3      : {index: 3, fn:['PWM' ,'ADC']},
        A4      : {index: 4, fn:['PWM' ,'ADC']},
        A5      : {index: 5, fn:['PWM' ,'ADC']},
        A6      : {index: 6, fn:['PWM' ,'ADC']},
        A7      : {index: 7, fn:['PWM' ,'ADC']},
        A8      : {index: 8, fn:['PWM'       ]},
        A9      : {index: 9, fn:['PWM'       ]},
        A10     : {index:10},
        A11     : {index:11},
        A12     : {index:12},
        A13     : {index:13},
        A14     : {index:14},
        A15     : {index:15},
        A16     : {index:16},
        A17     : {index:17},
        A18     : {index:18},
        A19     : {index:19},
        A20     : {index:20},
        A21     : {index:21},
        A22     : {index:22},
        A23     : {index:23},
        A24     : {index:24},
        A25     : {index:25},
        A26     : {index:26},
        A27     : {index:27},
        A28     : {index:28},
        A29     : {index:29},
        A30     : {index:30},
        A31     : {index:31},
        B0      : {index:32, fn:['PWM','MISO']},
        B1      : {index:33, fn:['PWM','SCLK']},
        B2      : {index:34, fn:['PWM','MOSI']},
        B3      : {index:35, fn:['PWM','SCL' ]},
        B4      : {index:36, fn:['PWM','SDA' ]},
        B5      : {index:37, fn:['PWMX'      ]},
        B6      : {index:38, fn:['PWMX'      ]},
        B7      : {index:39, fn:['PWMX'      ]},
        B8      : {index:40},
        B9      : {index:41},
        B10     : {index:42},
        B11     : {index:43},
        B12     : {index:44},
        B13     : {index:45},
        B14     : {index:46},
        B15     : {index:47},
        B16     : {index:48},
        B17     : {index:49},
        B18     : {index:50},
        B19     : {index:51},
        B20     : {index:52},
        B21     : {index:53},
        B22     : {index:54},
        B23     : {index:55},
        B24     : {index:56},
        B25     : {index:57},
        B26     : {index:58},
        B27     : {index:59},
        B28     : {index:60},
        B29     : {index:61},
        B30     : {index:62},
        B31     : {index:63},
        C0      : {index:64, fn:['PWM' ]},
        C1      : {index:65, fn:['PWM' ]},
        C2      : {index:66, fn:['PWM' ]},
        C3      : {index:67, fn:['PWM' ]},
        C4      : {index:68, fn:['PWM' ]},
        C5      : {index:69, fn:['PWM' ]},
        C6      : {index:70, fn:['PWM' ]},
        C7      : {index:71, fn:['PWM' ]},
        C8      : {index:72, fn:['PWM' ]},
        C9      : {index:73, fn:['PWM' ]},
        C10     : {index:74, fn:['PWM' ]},
        C11     : {index:75, fn:['PWM' ]},
        C12     : {index:76, fn:['PWM' ]},
        C13     : {index:77, fn:['PWM' ]},
        C14     : {index:78},
        C15     : {index:79},
        C16     : {index:80},
        C17     : {index:81},
        C18     : {index:82},
        C19     : {index:83},
        C20     : {index:84},
        C21     : {index:85},
        C22     : {index:86},
        C23     : {index:87},
        C24     : {index:88},
        C25     : {index:89},
        C26     : {index:90},
        C27     : {index:91},
        C28     : {index:92},
        C29     : {index:93},
        C30     : {index:94},
        C31     : {index:95}
    };
    //..........................................................................
    function processPin(pinDefinition, key)
    {
        const pinRecord = {};
        //......................................................................
        function processPinFunction(supports)
        {
            if (supports.slice(0,3) === 'ADC' ) pinRecord.hasADC  = true;
            if (supports.slice(0,3) === 'DAC' ) pinRecord.hasDAC  = true;
            if (supports.slice(0,3) === 'PWM' ) pinRecord.hasPWM  = true;
            if (supports            === 'PWMX') pinRecord.hasPWMX = true;
        }
        ////////////////////////////////////////////////////////////////////////
        //If pin has no listed functions then assume its just a GPIO, no specials.
        var pinFunctions;
        if (pinDefinition.fn)
        {
            pinFunctions = pinDefinition.fn;
            processList(pinFunctions,processPinFunction);

            pinFunctions.unshift(key);
            pinRecord.id           = key;
            pinRecord.caption      = pinFunctions.join('/');
            pinRecord.inputChanged = false;
            pinRecord.value        = 0;
            pinRecord.report       = false;
            pinRecord.index        = pinDefinition.index;

            gpio[pinRecord.id] = pinRecord;
        }
    }
    ////////////////////////////////////////////////////////////////////////////
    processList(pinInfo, processPin);
}
//..............................................................................
buildPinTemplates();

addToNameSpace(gpio, 'gpio');
module.exports = {gpio};
