// mesh-base/system-metadata/unit-type-groups.js
//..............................................................................
const {util} = cxq;
//..............................................................................

//..............................................................................
const
{
    processList,
} = util;
//..............................................................................

//..............................................................................
const unitTypeGroups =
{
    relative         : {caption_en : 'Relative Percentages'    , caption_zh : '相对百分比'},
    concentration    : {caption_en : 'Concentration'           , caption_zh : '浓度'},
    power            : {caption_en : 'Power'                   , caption_zh : '功率'},
    current          : {caption_en : 'Current'                 , caption_zh : '电流'},
    voltage          : {caption_en : 'Voltage'                 , caption_zh : '电压'},
    resistance       : {caption_en : 'Electrical Resistance'   , caption_zh : '电阻'},
    capacitance      : {caption_en : 'Electrical Capacitance'  , caption_zh : '电容'},
    inductance       : {caption_en : 'Electrical Inductance'   , caption_zh : '电感'},
    charge           : {caption_en : 'Electric Charge'         , caption_zh : '电量'},
    conductance      : {caption_en : 'Electrical Conductance'  , caption_zh : '电导'},
    magnetic_flux    : {caption_en : 'Magnetic Flux'           , caption_zh : '磁通量'},
    magnetic_field   : {caption_en : 'Magnetic Field Strength' , caption_zh : '磁场强度'},
    frequency        : {caption_en : 'Frequency'               , caption_zh : '频率'},
    data_flow_rate   : {caption_en : 'Data-Flow Rate'          , caption_zh : '数据流速率' },
    storage          : {caption_en : 'Storage'                 , caption_zh : '存储容量'},
    other_electrical : {caption_en : 'Other Electrical'        , caption_zh : '其他电气'},
    temperature      : {caption_en : 'Temperature'             , caption_zh : '温度'},
    humidity         : {caption_en : 'Humidity'                , caption_zh : '湿度'},
    pressure         : {caption_en : 'Pressure'                , caption_zh : '压强'},
    light            : {caption_en : 'Light'                   , caption_zh : '光'},
    sound            : {caption_en : 'Sound'                   , caption_zh : '声'},
    angles           : {caption_en : 'Angles'                  , caption_zh : '角'},
    force            : {caption_en : 'Force'                   , caption_zh : '力'},
    distance         : {caption_en : 'Distance'                , caption_zh : '距离'},
    speed            : {caption_en : 'Speed'                   , caption_zh : '速度'},
    rotation         : {caption_en : 'Rotation Speed'          , caption_zh : '转速'},
    acceleration     : {caption_en : 'Acceleration'            , caption_zh : '加速度'},
    time_span        : {caption_en : 'Time-Span'               , caption_zh : '时间跨度'}
};
//..............................................................................

//..............................................................................
processList(unitTypeGroups, function(group, id)
{
    group.id = id;
});
//..............................................................................

module.exports = unitTypeGroups;
