// mesh-base/system-metadata/package-types.js
//..............................................................................
const packageTypes =
{
    //..........................................................................
    // Software packages
    // - app means a package that runs on an execution platform
    // - rtm means runtime module, loaded by an app
    //..........................................................................
    app_browser     : {caption_en:'Browser App'       , caption_zh:'浏览器应用'},
    app_devos       : {caption_en:'Device OS'         , caption_zh:'设备系统'},
    app_edge        : {caption_en:'Device App'        , caption_zh:'设备应用'},
    app_mobile      : {caption_en:'Mobile App'        , caption_zh:'移动应用'},
    app_module      : {caption_en:'Module Firmware'   , caption_zh:'模块固件'},
    app_native      : {caption_en:'Native App'        , caption_zh:'原生应用'},
    app_node        : {caption_en:'Node App'          , caption_zh:'Node应用'},
    rtm_browser     : {caption_en:'Browser Plugin'    , caption_zh:'浏览器插件'},
    rtm_driver      : {caption_en:'Device Driver'     , caption_zh:'设备驱动'},
    rtm_native      : {caption_en:'Native Plugin'     , caption_zh:'原生插件'},
    rtm_node        : {caption_en:'Node Plugin'       , caption_zh:'Node插件'},
    //..........................................................................

    //..........................................................................
    // Hardware Packages
    //..........................................................................
    asm_pcb : {caption_en : 'PCB Assembly Package'    , caption_zh:'PCB组装包'},
    fab_pcb : {caption_en : 'PCB Fabrication Package' , caption_zh:'PCB生产包'},
    //..........................................................................

    //..........................................................................
    // Not currently used - may be used in the future
    //..........................................................................
    eda_symbol      : {caption_en : 'Schematic Symbol'             , caption_zh : '原理图符号'},
    eda_footprint   : {caption_en : 'PCB Footprint'                , caption_zh : 'PCB管脚图'},
    eda_schematic   : {caption_en : 'Schematic Sheet'              , caption_zh : '原理图设计图'},

    //..........................................................................
    // TODO NM: Remove in future remove, keep alive for now
    //..........................................................................
    eda_component   : {caption_en : 'EDA Component'                , caption_zh : 'EDA组件'},     // Replaced by component item
    sub_assembly    : {caption_en : 'Sub-Assembly'                 , caption_zh : '子组装'},      // Replaced by sub-assembly item
    snapshot        : {caption_en : 'Project Snapshot'             , caption_zh : '项目快照'},    // Replaced by component revision
    item_revision   : {caption_en : 'Item Revision'                , caption_zh : '条目更新'},    // Replaced by item revision
    bot             : {caption_en:  'Machine App'                  , caption_zh:'主机应用'},
    capability      : {caption_en:  'Machine Capability'           , caption_zh:'主机能力包'},
    app             : {caption_en:  'Browser Application'          , caption_zh:'浏览器应用'},
    plugin          : {caption_en:  'Browser App Plugin'           , caption_zh:'浏览器应用插件'},
    utility         : {caption_en:  'Machine Utility Package'      , caption_zh:'主机工具包'},
    driver          : {caption_en:  'Device Driver'                , caption_zh:'设备驱动包'},
    firmware        : {caption_en:  'Device Firmware Package'      , caption_zh:'设备固件包'},
};
//..............................................................................
module.exports = packageTypes;
