// mesh-base/util/color-serialization.js
var colorspaces = require('colorspaces');

//..............................................................................
const {logger, util} = cxq;
//..............................................................................

//..............................................................................
const
{
    isString,
    replaceAll,
} = util;
//..............................................................................

//..............................................................................
var log = logger('color-serialization');
log.setLevel('info');
//..............................................................................

//..............................................................................
const colorspace_CIEXYZ_To_hex  = colorspaces.converter('CIEXYZ', 'hex'   );
const colorspace_hex_To_CIEXYZ  = colorspaces.converter('hex'   , 'CIEXYZ');
//..............................................................................

//..............................................................................
function CIEXYZString_To_Hex(str)
{
    str = str.slice(7,-1).split(',');

    var x = parseFloat(str[0]);
    var y = parseFloat(str[1]);
    var z = parseFloat(str[2]);

    var hex = CIEXYZ_To_Hex({x,y,z});

    return hex;
}
//..............................................................................

//..............................................................................
function int24Color_To_Rgb(int)
{
    var red   = (int >> 16) & 0x0ff;
    var green = (int >>  8) & 0x0ff;
    var blue  = (int)       & 0x0ff;

    return {r:red, g:green, b:blue};
}
//..............................................................................

//..............................................................................
function CIEXYZ_To_Hex({x,y,z})
{
    // Returns "#FF00FF"
    return colorspace_CIEXYZ_To_hex([x,y,z]);
}
//..............................................................................

//..............................................................................
function hex_To_Int24(hex)
{
    if (isString(hex))
        hex = replaceAll(hex, "#", "");

    // Returns 3215
    var int = parseInt(hex, 16);
    return int24Color_To_Rgb(int);
}
//..............................................................................

//..............................................................................
function CIEXYZ_To_Rgb({x,y,z})
{
    // Returns {r:244, g:12, b:0}
    var hex = CIEXYZ_To_Hex({x,y,z});
    var int = hex_To_Int24(hex);
    var rgb = int24Color_To_Rgb(int);
    return int;
}
//..............................................................................

//..............................................................................
function hex_To_CIEXYZ(hex)
{
    // Returns {x:0.2323, y:0.323323, z:02212}
    var ciexyz = colorspace_hex_To_CIEXYZ(hex);
    return {x:ciexyz[0], y:ciexyz[1], z:ciexyz[2]};
}
//..............................................................................

module.exports =
{
    CIEXYZ_String:
    {
        toHex       : CIEXYZString_To_Hex
    },
    CIEXYZ:
    {
        toHex       : CIEXYZ_To_Hex,
        toRgb       : CIEXYZ_To_Rgb,
    },
    int24:
    {
        toRgb       : int24Color_To_Rgb
    },
    hex:
    {
        toInt24     : hex_To_Int24,
        toCIEXYZ    : hex_To_CIEXYZ
    }
}
