// mesh-base/util/qr-reader-factory.js
const jsqr = require('jsqr');
const imageFilters = require('./qr-reader-image-processors');

//..............................................................................
function createQrScanner(Buf, readImage)
{
    //..........................................................................
    function applyContrast(imageData)
    {
        imageFilters.contrast(imageData, 60);
        return imageData;
    }
    //..........................................................................
    function applyBlur(imageData)
    {        
        imageFilters.blur(imageData, 1);
        return imageData;
    }
    //..........................................................................
    function applyThreshold(imageData)
    {
        imageFilters.threshold(imageData, 150);
        return imageData;
    }
    //..........................................................................
    function applyResize(srcImageData, size)
    {
        const {width, height} = srcImageData;

        const dstWidth     = size;
        const dstHeight    = Math.floor(height * (dstWidth / width));
        const dstImageData = createImageData(dstWidth, dstHeight);

        imageFilters.bilinearInterpolation(srcImageData, dstImageData);
        
        return dstImageData;
    }
    //..........................................................................
    function createImageData(width, height)
    {
        return {data: new Buf(width*height*4), width, height};
    }
    //..........................................................................
    function cloneImageData(srcImageData)
    {   
        const target = 
        {
            data  : new Buf(srcImageData.width*srcImageData.height*4),
            width : srcImageData.width,
            height: srcImageData.height
        }

        for(var c=0; c < srcImageData.data.length; c++)
        {
            target.data[c] = srcImageData.data[c];
        }

        return target;
    }
    //..........................................................................
    function getQrDataFromImageData(imageData)
    {
        return jsqr(imageData.data, imageData.width, imageData.height);
    }
    //..........................................................................
    function detectQrCode(imageData)
    {
        var resizedImageData;
        
        var   qrData = null;
        const sizes  = [640, 1024];

        for (var c=0; c < sizes.length; c++)
        {
            var resizeWidth = sizes[c];

            if (imageData.width > resizeWidth)
                resizedImageData = applyResize(cloneImageData(imageData), resizeWidth);
            else
                resizedImageData = imageData;

            qrData = detectQrCodeWithImageFilters(resizedImageData);

            if (qrData.data)
                break;
        }
        
        return qrData;
    }
    //..........................................................................
    function detectQrCodeWithImageFilters(imageData)
    {
        var qrData         = null;
        var appliedFilters = [];
        var backupImage    = cloneImageData(imageData);

        qrData = getQrDataFromImageData(imageData);

        if (!qrData)
        {                
            imageData = applyBlur(imageData);
            appliedFilters.push('blur');
            qrData = getQrDataFromImageData(imageData);
        }

        if (!qrData)
        {
            imageData = cloneImageData(backupImage);
            imageData = applyContrast(imageData);
            appliedFilters.push('contrast');            
            qrData = getQrDataFromImageData(imageData);
        }

        if (!qrData)
        {
            imageData = cloneImageData(backupImage);
            imageData = applyThreshold(imageData);
            appliedFilters.push('threshold');            
            qrData = getQrDataFromImageData(imageData);
        }

        if (!qrData)
        {
            imageData = cloneImageData(backupImage);
            imageData = applyContrast(imageData);
            imageData = applyBlur(imageData);
            imageData = applyThreshold(imageData);
            appliedFilters.push('contrast+blur+threshold');            
            qrData = getQrDataFromImageData(imageData);
        }

        const result = 
        {
            imageData,
            appliedFilters,
            width   : imageData.width,
            height  : imageData.height,
            data    : null,
            location: null,
        };

        if (qrData)
        {
            result.data     = qrData.data;
            result.location = qrData.location;
        }

        return result;
    }
    //..........................................................................    
    function publicScope(source, callback)
    {
        function callback_readImage(err, imageData)
        {
            var result;
            
            if (err)
                return callback(err);

            try
            {
                result = detectQrCode(imageData);
            }
            catch(e)
            {
                err = e;
                result = null;
            }

            callback(err, result);
        }
        //......................................................................        
        readImage(source, callback_readImage);            
    }
    //..........................................................................
    return publicScope;
}
//..............................................................................

module.exports = createQrScanner;
