// mesh-doc-gen/static/ds-constants.js
//..............................................................................

const{colorsHex    } = cxq.meshDocGen;

const
{
    util,
} = cxq;

const {addProperties} = util;

//..............................................................................

//..............................................................................
const headerDefs =
[
    '<defs>',
    '        <pattern',
    '            id    ="WHITE-SQUARES"',
    '            x     ="10"',
    '            y     ="10"',
    '            width ="40"',
    '            height="40"',
    '            patternUnits="userSpaceOnUse"',
    '            >',
    '            <rect x="00" y="00" width="40" height="40" fill="' + colorsHex.colorSchGridGrey + '" />',
    '',
    '            <rect x="10" y="10" width="10" height="10" fill="' + colorsHex.colorWhite + '" />',
    '            <rect x="30" y="10" width="10" height="10" fill="' + colorsHex.colorWhite + '" />',
    '            <rect x="10" y="30" width="10" height="10" fill="' + colorsHex.colorWhite + '" />',
    '            <rect x="30" y="30" width="10" height="10" fill="' + colorsHex.colorWhite + '" />',
    '        </pattern>',
    '',
    '    <!-- .................................................................  -->',
    '    <!-- Ports  -->',
    '    <!-- .................................................................  -->',
    '    <svg',
    '        id          ="PORT-INPUT"',
    '        overflow    ="visible"',
    '        stroke      ="' + colorsHex.colorStroke + '"',
    '        fill        ="' + colorsHex.colorSchPortFill + '"',
    '        stroke-width="0.1"',
    '        shape-rendering="crispEdges"'    ,
    '        >',
    '        <path d="m 0, 0 h 95 l 5, 5 l -5, 5 h -95 v -10"/>',
    '    </svg>',
    '    <svg',
    '        id          ="PORT-OUTPUT"',
    '        overflow    ="visible"',
    '        stroke      ="' + colorsHex.colorStroke + '"',
    '        fill        ="' + colorsHex.colorSchPortFill + '"',
    '        shape-rendering="crispEdges"'    ,
    '        stroke-width="0.1"',
    '        >',
    '        <path d="m 5, 0 h 95 v 10 h -95 l -5, -5 l 5, -5 "/>',
    '    </svg>',
    '    <svg',
    '        id          ="PORT-BI-DIRECTIONAL"',
    '        overflow    ="visible"',
    '        stroke      ="' + colorsHex.colorStroke + '"',
    '        fill        ="' + colorsHex.colorSchPortFill + '"',
    '        shape-rendering="crispEdges"'    ,
    '        stroke-width="0.1"',
    '        >',
        '        <path d="m 5, 0 h 90 l 5, 5 l -5, 5 h -90 l -5, -5 z"/>',
    '    </svg>',
    '    <svg',
    '        id          ="PORT-PASSIVE"',
    '        overflow    ="visible"',
    '        stroke      ="' + colorsHex.colorStroke + '"',
    '        fill        ="' + colorsHex.colorSchPortFill + '"',
    '        shape-rendering="crispEdges"'    ,
    '        stroke-width="0.1"',
    '        >',
    '        <path d="m 0, 0 h 100 v 10 h -100 z"/>',
    '    </svg>',
    '    <!-- .................................................................  -->',
    '    <!-- Pins  -->',
    '    <!-- .................................................................  -->',
    '    <svg',
    '        id          ="PIN-INPUT"',
    '        overflow    ="visible"',
    '        stroke      ="' + colorsHex.colorStroke + '"',
    '        fill        ="' + colorsHex.colorWhite + '"',
    '        shape-rendering="crispEdges"'    ,
    '        stroke-width="0.75"',
    '        >',
    '        <line x1="0" x2="+15" y1="0" y2="0"/>',
    '        <path d="m 14, -2 l 6, 2 l -6, 2 z" stroke-width="0.2"/>',
    '    </svg>',
    '    <svg',
    '        id          ="PIN-OUTPUT"',
    '        overflow    ="visible"',
    '        stroke      ="' + colorsHex.colorStroke + '"',
    '        fill        ="' + colorsHex.colorWhite + '"',
    '        shape-rendering="crispEdges"'    ,
    '        stroke-width="0.75"',
    '        >',
    '        <line x1="0" x2="+15" y1="0" y2="0"/>',
    '        <path d="m 14, 0 l 6, -2 v 4 z" stroke-width="0.2"/>',
    '    </svg>',
    '    <svg',
    '        id          ="PIN-BI-DIRECTIONAL"',
    '        overflow    ="visible"',
    '        stroke      ="' + colorsHex.colorStroke + '"',
    '        fill        ="' + colorsHex.colorWhite + '"',
    '        shape-rendering="crispEdges"'    ,
    '        stroke-width="0.75"',
    '        >',
    '        <line x1="0" x2="+15" y1="0" y2="0"/>',
    '        <path d="m 14, -2 l 6, +2 l -6, 2 z" stroke-width="0.2"/>',
    '        <path d="m  7, +0 l 6, -2 v +4    z" stroke-width="0.2"/>',
    '    </svg>',
    '    <svg',
    '        id          ="PIN-PASSIVE"',
    '        overflow    ="visible"',
    '        stroke      ="' + colorsHex.colorStroke + '"',
    '        fill        ="' + colorsHex.colorWhite + '"',
    '        shape-rendering="crispEdges"'    ,
    '        stroke-width="0.75"',
    '        >',
    '        <line x1="0" x2="+20" y1="0" y2="0"/>',
    '    </svg>',
    '    <svg',
    '        id          ="PIN-PASSIVE-V"',
    '        overflow    ="visible"',
    '        stroke      ="' + colorsHex.colorStroke + '"',
    '        fill        ="' + colorsHex.colorWhite + '"',
    '        shape-rendering="crispEdges"'    ,
    '        stroke-width="0.75"',
    '        >',
    '        <line x1="0" x2="0" y1="0" y2="20"/>',
    '    </svg>',
    '    <svg',
    '        id          ="PIN-3V3"',
    '        overflow    ="visible"',
    '        stroke      ="' + colorsHex.color3V3 + '"',
    '        shape-rendering="geometricPrecision"',
    '        stroke-width="3"',
    '        >',
    '        <line x1="-10" x2="0" y1="10" y2="10" stroke-linecap="round" stroke-linejoin="round"/>',
    '    </svg>',
    '    <svg',
    '        id          ="PIN-GND"',
    '        overflow    ="visible"',
    '        stroke      ="' + colorsHex.colorGround + '"',
    '        shape-rendering="geometricPrecision"',
    '        stroke-width="3"',
    '        >',
    '        <line x1="-10" x2="+00" y1="10" y2="10" stroke-linecap="round" stroke-linejoin="round"/>',
    '    </svg>',
    '    <svg',
    '        id          ="PIN-GND-V"',
    '        overflow    ="visible"',
    '        stroke      ="' + colorsHex.colorGround + '"',
    '        shape-rendering="geometricPrecision"',
    '        stroke-width="3"',
    '        >',
    '        <line x1="0" x2="0" y1="0" y2="10" stroke-linecap="round" stroke-linejoin="round"/>',
    '    </svg>',
    '    <!-- .................................................................  -->',
    '    <svg',
    '        id          ="PIN-POWER"',
    '        overflow    ="visible"',
    '        stroke      ="' + colorsHex.colorSchPinPower + '"',
    '        shape-rendering="geometricPrecision"',
    '        stroke-width="3"',
    '        >',
    '        <line x1="00" x2="10" y1="10" y2="10" stroke-linecap="round" stroke-linejoin="round"/>',
    '    </svg>',
    '    <!-- .................................................................  -->',
    '    <!-- Wires & Junction  -->',
    '    <!-- .................................................................  -->',
    '    <svg',
    '        id          ="WIRE-PLAIN"',
    '        overflow    ="visible"',
    '        stroke      ="' + colorsHex.colorSchWire + '"',
    '        shape-rendering="crispEdges"'    ,
    '        stroke-width="0.75"',
    '        >',
    '        <line x1="0" x2="+50" y1="0" y2="0"/>',
    '    </svg>',
    '    <svg',
    '        id             ="JUNCTION"',
    '        overflow       ="visible"',
    '        fill           ="' + colorsHex.colorSchJunction + '"',
    '        shape-rendering="geometricPrecision"',
    '        >',
    '        <circle cx="0" cy="0" r="2.5"/>',
    '    </svg>',
    ''
];
//..........................................................................
const footerDefs = '</defs>';
//..........................................................................
const headerBody =
[
    '    <rect width="1133" height="760" fill="url(#WHITE-SQUARES)" stroke="' + colorsHex.colorStroke + '" stroke-width="0.5"/>',
    '',
];
//..............................................................................

//..............................................................................
const headerRoot =
[
    '<svg                                           ',
    '    xmlns        ="http://www.w3.org/2000/svg" ',
    '    id           ="device-sheet"'               ,
    '    font-family  ="consolas"'                   ,
    '    font-size    ="0.22"'                       ,
    '    cursor       ="default"'                    ,
    '    language     ="language_en"'                ,
    '    viewBox      ="0 0 1133 760"'               ,
    '    width        ="1133"'                       ,
    '    height       ="760"'                        ,
    '    >'                                          ,
];
//..............................................................................

//..............................................................................
const footerRoot = '</svg>';
//..............................................................................

//..............................................................................
const systemPowerNets =
[
    'GND',
    '3V3',
    '3V3_IN',
    '3V3_OUT',
    '5V0',
    '5V',
    '5V0_IN',
    '5V0_OUT',
    '5V_IN',
    '5V_OUT',
    '9V_OUT',
    'VBAT',
    'VA_IN',
    'VB_IN',
];
//..............................................................................

//..............................................................................
const fontSizesDS =
{
    fontSizeDeviceSheet : 9,
    fontSizeSchPart     : 10,
    fontSizeStandardMB  : 0.22,
};
//..............................................................................

//..............................................................................
const fontsDS =
{
    fontPort        : 'Times New Roman',
    fontSchPart     : 'Times New Roman',
    fontLabel       : 'Arial',
    fontTitle       : 'Arial',
    fontDescription : 'Arial',
    fontTextAdv     : 'Arial',
};
//..............................................................................

//..............................................................................
const pinTypesPower =
[
    'ground',
    'power_consumer',
    'power_provider',
];
//..............................................................................

//..............................................................................
const specialPinFunctions =
[
    'RESET',
    'SPI',
    'SPI_MCU',
    'I2C',
    'I2C_MCU',
    'SW_DEBUG',
    'WAKE_IN_OUT',
];
//..............................................................................

//..............................................................................
const exported =
{
    fontSizesDS,
    fontsDS,
    footerDefs,
    footerRoot,
    headerBody,
    headerDefs,
    headerRoot,
    pinTypesPower,
    specialPinFunctions,
    systemPowerNets,
};

module.exports = exported;
